# Applied Data Science Capstone

## Segmenting and Clustering Neighborhoods in Toronto

## Week 3 - Part 1

To read HTML webpages into pandas \
    - import pandas as pd \
    - use pd.read_html() function to read HTML webpage

In [1]:
import pandas as pd
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
df= pd.read_html(url, header = 0)

In [2]:
df_canada=df[0]
df_canada.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Only the cells that have an assigned borough are processed. So the cells with a borough that is 'Not assigned' will be dropped.

In [3]:
import numpy as np
df_canada["Borough"].replace('Not assigned', np.nan, inplace=True)
df_canada.dropna(subset=["Borough"], axis=0, inplace=True)
df_canada.reset_index(drop=True, inplace=True)
df_canada.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

More than one neighborhood can exist in one postal code area. These  rows will be combined into one row with the neighborhoods separated with a comma.

In [4]:
df_canada.shape

(103, 3)

In [5]:
df_canada["Postal Code"].value_counts

<bound method IndexOpsMixin.value_counts of 0      M3A
1      M4A
2      M5A
3      M6A
4      M7A
      ... 
98     M8X
99     M4Y
100    M7Y
101    M8Y
102    M8Z
Name: Postal Code, Length: 103, dtype: object>

The number of raws and length of value_counts of Postal Codes are equal(103). So there aren't any postal code area in which more than one neighbourhood exist.

If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough

In [6]:
df_canada["Neighbourhood"].replace('Not assigned', df_canada["Borough"], inplace=True)
df_canada

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [7]:
df_canada.shape

(103, 3)

## Week 3 - Part 2

Next is building a dataframe of the postal code, borough name, neighborhood name, latitude and longitude coordinates of each neighborhood. The csv file is used to create the following dataframe of the geographical coordinates of the neighborhoods.

In [8]:
import pandas as pd
filename = 'https://cocl.us/Geospatial_data'
df_loc = pd.read_csv(filename)

df_loc

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

Then, two dataframes are merged using inner join on "Postal Code" column.

In [9]:
df_canada=pd.merge(df_canada,df_loc,on='Postal Code')
df_canada

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

## Week 3 - Part 3

Exploring and clustering the neighborhoods in Toronto

In [10]:
import numpy as np 

import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json 

!pip install geopy
from geopy.geocoders import Nominatim 

import requests 
from pandas.io.json import json_normalize 

import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

!pip install folium
import folium

print('Libraries imported.')

Libraries imported.


#### Simplify the segment and cluster

Let's simplify the segment and cluster only the neighborhoods in Toronto. So let's slice the original dataframe and create a new dataframe of the boroughs which contain the word 'Toronto'.

In [46]:
toronto_data = df_canada[df_canada['Borough'].str.contains("Toronto")].reset_index(drop=True)
toronto_data.head()

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

Let's get the geographical coordinates of Toronto.

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them. Firstly, let's start with defining Foursquare Credentials and Version.

In [47]:
CLIENT_ID = 'FKCGWGBAY4RGGYB02SCBRLOSDN4CWUANFY5ODSQ3YLK1ZCA5' # your Foursquare ID
CLIENT_SECRET = 'DAPEQAYLFNGU03NXGA45LQN1M353TXMZ01KUZRE2HGD2KMFI' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: FKCGWGBAY4RGGYB02SCBRLOSDN4CWUANFY5ODSQ3YLK1ZCA5
CLIENT_SECRET:DAPEQAYLFNGU03NXGA45LQN1M353TXMZ01KUZRE2HGD2KMFI


Now let's create a function to get the top 100 venues within a radius of 500 meters for all the neighborhoods in Toronto.

In [121]:
def getNearbyVenues(names, latitudes, longitudes, LIMIT=100, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [122]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighbourhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude'])

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

In [123]:
toronto_venues.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Regent Park, Harbourfront                43.65426               -79.360636   
1  Regent Park, Harbourfront                43.65426               -79.360636   
2  Regent Park, Harbourfront                43.65426               -79.360636   
3  Regent Park, Harbourfront                43.65426               -79.360636   
4  Regent Park, Harbourfront                43.65426               -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3          Impact Kitchen       43.656369       -79.356980   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3           Restaurant  
4                  Spa

#### Analyze Each Neighborhood

Let's start to analyze each neighbourhood

In [124]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[toronto_onehot.columns.get_loc('Neighbourhood')]] + list(toronto_onehot[toronto_onehot.columns.difference(['Neighbourhood'])])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head(20)

Neighbourhood  Afghan Restaurant  Airport  Airport Food Court  \
0   Regent Park, Harbourfront                  0        0                   0   
1   Regent Park, Harbourfront                  0        0                   0   
2   Regent Park, Harbourfront                  0        0                   0   
3   Regent Park, Harbourfront                  0        0                   0   
4   Regent Park, Harbourfront                  0        0                   0   
5   Regent Park, Harbourfront                  0        0                   0   
6   Regent Park, Harbourfront                  0        0                   0   
7   Regent Park, Harbourfront                  0        0                   0   
8   Regent Park, Harbourfront                  0        0                   0   
9   Regent Park, Harbourfront                  0        0                   0   
10  Regent Park, Harbourfront                  0        0                   0   
11  Regent Park, Harbourfront                  0        0                   0   
12  Regent Park, Harbourfront                  0        0                   0   
13  Regent Park, Harbourfront                  0        0                   0   
14  Regent Park, Harbourfront                  0        0                   0   
15  Regent Park, Harbourfront                  0        0                   0   
16  Regent Park, Harbourfront                  0        0                   0   
17  Regent Park, Harbourfront                  0        0                   0   
18  Regent Park, Harbourfront                  0        0                   0   
19  Regent Park, Harbourfront                  0        0                   0   

    Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0              0               0                0                 0   
1              0               0                0                 0   
2              0               0                0                 0   
3              0               0                0                 0   
4              0               0                0                 0   
5              0               0                0                 0   
6              0               0                0                 0   
7              0               0                0                 0   
8              0               0                0                 0   
9              0               0                0                 0   
10             0               0                0                 0   
11             0               0                0                 0   
12             0               0                0                 0   
13             0               0                0                 0   
14             0               0                0                 0   
15             0               0                0                 0   
16             0               0                0                 0   
17             0               0                0                 0   
18             0               0                0                 0   
19             0               0                0                 0   

    American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                     0             0         0            0           0   
1                     0             0         0            0           0   
2                     0             0         0            0           0   
3                     0             0         0            0           0   
4                     0             0         0            0           0   
5                     0             0         0            0           0   
6                     0             0         0            0           0   
7                     0             0         0            0           0   
8                     0             0         0            0           0   
9                     0             0         0            0  

In [113]:
toronto_onehot.shape

(1639, 240)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [125]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Afghan Restaurant  \
0                                         Berczy Park           0.000000   
1        Brockton, Parkdale Village, Exhibition Place           0.000000   
2   Business reply mail Processing Centre, South C...           0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
4                                  Central Bay Street           0.000000   
5                                            Christie           0.000000   
6                                Church and Wellesley           0.013333   
7                      Commerce Court, Victoria Hotel           0.000000   
8                                          Davisville           0.000000   
9                                    Davisville North           0.000000   
10                       Dufferin, Dovercourt Village           0.000000   
11             First Canadian Place, Underground city           0.000000   
12    Forest Hill North & West, Forest Hill Road Park           0.000000   
13                           Garden District, Ryerson           0.000000   
14  Harbourfront East, Union Station, Toronto Islands           0.000000   
15                      High Park, The Junction South           0.000000   
16                     India Bazaar, The Beaches West           0.000000   
17          Kensington Market, Chinatown, Grange Park           0.000000   
18                                      Lawrence Park           0.000000   
19                           Little Portugal, Trinity           0.000000   
20                        Moore Park, Summerhill East           0.000000   
21                  North Toronto West, Lawrence Park           0.000000   
22                             Parkdale, Roncesvalles           0.000000   
23        Queen's Park, Ontario Provincial Government           0.000000   
24                          Regent Park, Harbourfront           0.000000   
25                           Richmond, Adelaide, King           0.000000   
26                                           Rosedale           0.000000   
27                                           Roselawn           0.000000   
28                                 Runnymede, Swansea           0.000000   
29                                     St. James Town           0.000000   
30                        St. James Town, Cabbagetown           0.000000   
31                                     Stn A PO Boxes           0.000000   
32                                    Studio District           0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...           0.000000   
34                The Annex, North Midtown, Yorkville           0.000000   
35                                        The Beaches           0.000000   
36                       The Danforth West, Riverdale           0.000000   
37           Toronto Dominion Centre, Design Exchange           0.000000   
38                     University of Toronto, Harbord           0.000000   

     Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0   0.000000            0.000000      0.000000        0.000000   
1   0.000000            0.000000      0.000000        0.000000   
2   0.000000            0.000000      0.000000        0.000000   
3   0.058824            0.058824      0.058824        0.117647   
4   0.000000            0.000000      0.000000        0.000000   
5   0.000000            0.000000      0.000000        0.000000   
6   0.000000            0.000000      0.000000        0.000000   
7   0.000000            0.000000      0.000000        0.000000   
8   0.000000            0.000000      0.000000        0.000000   
9   0.000000            0.000000      0.000000        0.000000   
10  0.000000            0.000000      0.000000        0.000000   
11  0.000000            0.000000      0.000000        0.000000   
12  0.000000            0.000000      0.000000        0.000000   
13  0.000000            0.000000      0.000000        0.000000   
14  0.000

Now, first let's write a function to sort the venues in descending order and then create the new dataframe and display the top 10 venues for each neighborhood.

In [115]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [132]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...    Light Rail Station   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery          Cocktail Bar        Farmers Market   
1        Breakfast Spot             Nightclub           Coffee Shop   
2           Pizza Place               Brewery         Burrito Place   
3       Airport Service              Boutique                   Bar   
4                  Café        Sandwich Place    Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0              Beer Bar            Restaurant           Cheese Shop   
1                   Gym            Playground               Stadium   
2               Butcher        Farmers Market  Fast Food Restaurant   
3         Boat or Ferry      Sculpture Garden         Historic Site   
4      Department Store   Japanese Restaurant           Salad Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0    Seafood Restaurant          Concert Hall     Basketball Stadium  
1                   Bar    Italian Restaurant                 Bakery  
2            Restaurant      Recording Studio          Auto Workshop  
3   Rental Car Location       Harbor / Marina            Coffee Shop  
4       Bubble Tea Shop          Burger Joint        Thai Restaurant

#### Cluster Neighborhoods

Run k-means to cluster the neighborhood into 5 clusters.

In [133]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2], dtype=int32)

Now, create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [134]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() 

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               2           Coffee Shop   
1  43.662301 -79.389494               2           Coffee Shop   
2  43.657162 -79.378937               2           Coffee Shop   
3  43.651494 -79.375418               2           Coffee Shop   
4  43.676357 -79.293031               3                   Pub   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Park                Bakery                   Pub   
1           Yoga Studio                 Diner            Restaurant   
2        Clothing Store                  Café       Bubble Tea Shop   
3                  Café          Cocktail Bar            Restaurant   
4     Health Food Store                 Trail          Neighborhood   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0         Breakfast Spot                  Café               Theater   
1  Portuguese Restaurant                  Park           Music Venue   
2    Japanese Restaurant        Cosmetics Shop                 Hotel   
3               Beer Bar             Gastropub   American Restaurant   
4            Yoga Studio              Dive Bar    Dim Sum Restaurant   

  8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0            Shoe Store               Brewery                 Restaurant  
1    Mexican Restaurant    Italian Restaurant                 Hobby Shop  
2             Bookstore           Pizza Place  Middle Eastern Restaurant  
3        Farmers Market                 Hotel        Japanese Restaurant  
4                 Diner        Discount Store        Distribution Center

Finally, let's visualize the resulting clusters

In [135]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Now, let's examine each cluster and determine the discriminating venue categories that distinguish each cluster.

In [143]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + [2] + list(range(5, toronto_merged.shape[1]))]]

Borough                                    Neighbourhood  \
18   Central Toronto                                    Lawrence Park   
21   Central Toronto  Forest Hill North & West, Forest Hill Road Park   
33  Downtown Toronto                                         Rosedale   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
18               0                  Park              Bus Line   
21               0      Sushi Restaurant                  Park   
33               0                  Park                 Trail   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
18           Swim School    Dim Sum Restaurant           Escape Room   
21    Mexican Restaurant         Jewelry Store                 Trail   
33          Tennis Court            Playground            Donut Shop   

   6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
18     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   
21              Dive Bar           Dim Sum Restaurant                 Diner   
33      Doner Restaurant                      Dog Run   Dumpling Restaurant   

   9th Most Common Venue 10th Most Common Venue  
18            Donut Shop       Doner Restaurant  
21        Discount Store    Distribution Center  
33         Deli / Bodega               Dive Bar

First cluster label which is reperesented with red color at the map looks like the neighbourhoods where the most common venues are parks

In [144]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1]+ [2] + list(range(5, toronto_merged.shape[1]))]]

Borough                Neighbourhood  Cluster Labels  \
29  Central Toronto  Moore Park, Summerhill East               1   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
29            Playground                 Trail           Yoga Studio   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
29          Dessert Shop           Escape Room     Electronics Store   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
29  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   10th Most Common Venue  
29       Doner Restaurant

Second cluster label which is reperesented with purple color at the map looks like the neighbourhoods where the most common venues are playgrounds

In [147]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1]+ [2] + list(range(5, toronto_merged.shape[1]))]]

Borough                                      Neighbourhood  \
0   Downtown Toronto                          Regent Park, Harbourfront   
1   Downtown Toronto        Queen's Park, Ontario Provincial Government   
2   Downtown Toronto                           Garden District, Ryerson   
3   Downtown Toronto                                     St. James Town   
5   Downtown Toronto                                        Berczy Park   
6   Downtown Toronto                                 Central Bay Street   
7   Downtown Toronto                                           Christie   
8   Downtown Toronto                           Richmond, Adelaide, King   
9       West Toronto                       Dufferin, Dovercourt Village   
10  Downtown Toronto  Harbourfront East, Union Station, Toronto Islands   
11      West Toronto                           Little Portugal, Trinity   
12      East Toronto                       The Danforth West, Riverdale   
13  Downtown Toronto           Toronto Dominion Centre, Design Exchange   
14      West Toronto       Brockton, Parkdale Village, Exhibition Place   
15      East Toronto                     India Bazaar, The Beaches West   
16  Downtown Toronto                     Commerce Court, Victoria Hotel   
17      East Toronto                                    Studio District   
20   Central Toronto                                   Davisville North   
22      West Toronto                      High Park, The Junction South   
23   Central Toronto                  North Toronto West, Lawrence Park   
24   Central Toronto                The Annex, North Midtown, Yorkville   
25      West Toronto                             Parkdale, Roncesvalles   
26   Central Toronto                                         Davisville   
27  Downtown Toronto                     University of Toronto, Harbord   
28      West Toronto                                 Runnymede, Swansea   
30  Downtown Toronto          Kensington Market, Chinatown, Grange Park   
31   Central Toronto  Summerhill West, Rathnelly, South Hill, Forest...   
32  Downtown Toronto  CN Tower, King and Spadina, Railway Lands, Har...   
34  Downtown Toronto                                     Stn A PO Boxes   
35  Downtown Toronto                        St. James Town, Cabbagetown   
36  Downtown Toronto             First Canadian Place, Underground city   
37  Downtown Toronto                               Church and Wellesley   
38      East Toronto  Business reply mail Processing Centre, South C...   

    Cluster Labels          1st Most Common Venue 2nd Most Common Venue  \
0                2                    Coffee Shop                  Park   
1                2                    Coffee Shop           Yoga Studio   
2                2                    Coffee Shop        Clothing Store   
3                2                    Coffee Shop                  Café   
5                2                    Coffee Shop                Bakery   
6                2                    Coffee Shop                  Café   
7                2                  Grocery Store                  Café   
8                2                    Coffee Shop                  Café   
9                2                       Pharmacy                Bakery   
10               2                    Coffee Shop              Aquarium   
11               2                            Bar           Coffee Shop   
12               2               Greek Restaurant           Coffee Shop   
13               2                    Coffee Shop                 Hotel   
14               2                           Café        Breakfast Spot   
15               2                           Park                   Gym   
16               2                    Coffee Shop            Restaurant   
17               2                    Coffee Shop                Bakery   
20               2               Department Store        Sandwich Place   
22               2             Mexica

Third cluster label which is reperesented with blue color at the map looks like the neighbourhoods where the most common venues are restaurants and coffee shops

In [145]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1]+ [2] + list(range(5, toronto_merged.shape[1]))]]

Borough Neighbourhood  Cluster Labels 1st Most Common Venue  \
4  East Toronto   The Beaches               3                   Pub   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4     Health Food Store                 Trail          Neighborhood   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
4           Yoga Studio              Dive Bar    Dim Sum Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
4                 Diner        Discount Store    Distribution Center

Fourth cluster label which is reperesented with green color at the map looks like the neighbourhoods where the most common venues are pubs and health food stores.

In [149]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1]+ [2] + list(range(5, toronto_merged.shape[1]))]]

Borough Neighbourhood  Cluster Labels 1st Most Common Venue  \
19  Central Toronto      Roselawn               4              Dive Bar   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19                Garden           Music Venue           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
19      Department Store     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19   Dumpling Restaurant            Donut Shop       Doner Restaurant

The last cluster label which is reperesented with yellow color at the map looks like the neighbourhoods where the most common venues are dive bars and gardens

Thanks for reviewing!